<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 3.2//EN">
<HTML>
 <HEAD>
<!-- Global site tag (gtag.js) - Google Analytics -->
<script async src="https://www.googletagmanager.com/gtag/js?id=UA-51353372-3"></script>
<script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-51353372-3');
</script>
   <TITLE> [nycphp-talk] PHP 5 Throwing Exceptions in Constructors
   </TITLE>
   <LINK REL="Index" HREF="http://nyphp.org/pipermail/talk/2005-February/index.html" >
   <LINK REL="made" HREF="mailto:talk%40lists.nyphp.org?Subject=Re:%20Re%3A%20%5Bnyphp-talk%5D%20PHP%205%20Throwing%20Exceptions%20in%20Constructors&In-Reply-To=%3C20050209221504.C368BA87B0%40virtu.nyphp.org%3E">
   <META NAME="robots" CONTENT="index,nofollow">
   <META http-equiv="Content-Type" content="text/html; charset=us-ascii">
   <LINK REL="Previous"  HREF="013596.html">
   <LINK REL="Next"  HREF="013600.html">
 <meta name="description" content="PHP jobs stackware web application support and development forums for Mongo LAMP MySQL SQL Windows IIS Linux." />
<meta name="author" content="Stackware Web Development - http://stackware.com" />
<meta name="verify-v1" content="zWMoPwMn/dolscLkjLXTpRqgr9K8bzlzL1vM9WTjpHc=" />
<meta name="msvalidate.01" content="F85F81C1D0BAB5C349E8395628845E27" />
<meta name="y_key" content="265502be3db4fb6e" />
</HEAD>
         <BODY BGCOLOR="#ffffff">
<h1 style="float: right; margin: 0; padding: 0 0 0 15px;"><a href="https://www.meetup.com/new-york-php/">NYCPHP Meetup</a></h1><h1><a href="http://www.nyphp.org">NYPHP.org</a></h1>
   <H1>[nycphp-talk] PHP 5 Throwing Exceptions in Constructors</H1>
    <B>Hans Zaunere</B> 
    <A HREF="mailto:talk%40lists.nyphp.org?Subject=Re:%20Re%3A%20%5Bnyphp-talk%5D%20PHP%205%20Throwing%20Exceptions%20in%20Constructors&In-Reply-To=%3C20050209221504.C368BA87B0%40virtu.nyphp.org%3E"
       TITLE="[nycphp-talk] PHP 5 Throwing Exceptions in Constructors">lists at zaunere.com
       </A><BR>
    <I>Wed Feb  9 17:15:00 EST 2005</I>
    <P><UL>
        <LI>Previous message: <A HREF="013596.html">[nycphp-talk] PHP 5 Throwing Exceptions in Constructors
</A></li>
        <LI>Next message: <A HREF="013600.html">[nycphp-talk] PHP 5 Throwing Exceptions in Constructors
</A></li>
         <LI> <B>Messages sorted by:</B> 
              <a href="date.html#13599">[ date ]</a>
              <a href="thread.html#13599">[ thread ]</a>
              <a href="subject.html#13599">[ subject ]</a>
              <a href="author.html#13599">[ author ]</a>
         </LI>
       </UL>
    <HR>  
<!--beginarticle-->
<PRE>
&gt;<i> &gt; All PHP 5 code examples I've seen show try/catch pairs like this:
</I>&gt;<i> &gt;
</I>&gt;<i> &gt; try {
</I>&gt;<i> &gt; 	$obj = new MyClass;
</I>&gt;<i> &gt; 	$obj-&gt;DoSomething();
</I>&gt;<i> &gt; } catch( Exception $e ) {
</I>&gt;<i> &gt; 	echo $e-&gt;getMessage();
</I>&gt;<i> &gt; }
</I>&gt;<i> &gt;
</I>&gt;<i> 
</I>&gt;<i> This is correct.
</I>&gt;<i> 
</I>&gt;<i> &gt; I am the only one in thinking that these examples are incorrect 99%
</I>&gt;<i> &gt; of the time?  Based on other languages, like Java, exceptions should
</I>&gt;<i> &gt; very very rarely be thrown from a class' constructor.  The simple
</I>&gt;<i> &gt; reason, as in PHP 5, is that the object never gets created if an
</I>&gt;<i> &gt; exception is thrown in the constructor.
</I>&gt;<i> 
</I>&gt;<i> PHP is not Java. :)
</I>
Riiiight... but it does have the same behavior :)

&gt;<i> &gt; Is there something particular to PHP 5 that I'm missing?  While it
</I>&gt;<i> &gt; could be considered pedantic, PHP 5 code examples should be as:
</I>&gt;<i> 
</I>&gt;<i> &gt; $obj = new MyClass;
</I>&gt;<i> &gt;
</I>&gt;<i> &gt; try {
</I>&gt;<i> &gt; 	$obj-&gt;DoSomething();
</I>&gt;<i> &gt; } catch( Exception $e ) {
</I>&gt;<i> &gt; 	echo $e-&gt;getMessage();
</I>&gt;<i> &gt; }
</I>&gt;<i> &gt;
</I>&gt;<i> 
</I>&gt;<i> This is incorrect.
</I>&gt;<i> 
</I>&gt;<i> &gt; In fact, in PHP 5 I cannot find anyway to determine the &quot;return&quot;
</I>&gt;<i> &gt; value of the new operator.  This leads me to believe that PHP 5's
</I>&gt;<i> &gt; concept of proper OO is to always assume the object is created
</I>&gt;<i> &gt; successfully, unless a catastrophic event happens, in which case an
</I>&gt;<i> &gt; exception is thrown, and the object isn't created at all.  This of
</I>&gt;<i> &gt; course makes sense, but then I'm confused as to why the code samples
</I>&gt;<i> &gt; all seem a bit incorrect.
</I>&gt;<i> 
</I>&gt;<i> In PHP 5, the way to signal failure in a constructor is to throw an
</I>&gt;<i> exception. There is no other way to communicate errors. As &quot;new&quot; will
</I>&gt;<i> not return &quot;false&quot;.
</I>&gt;<i> 
</I>&gt;<i> Exceptions do not have to occur due to catastrophic failures, such as
</I>&gt;<i> lack of memory. This can also occur due to bad input parameters. For
</I>&gt;<i> example, passing the wrong number of parameters, or if you try to
</I>&gt;<i> create a DOMElement with an illegal XML element name.
</I>
But what if a class should still be instantiated, even if some of the arguments are incorrect, or subsequent parts fail for some reason?  True, this validation could be done outside of the constructor, even in a totally different class that the constructor instantiates itself, but this still would prevent the instantiation of the original class.

class Top
{
	protected $Child;

	public function __construct( $value ) {
		$this-&gt;Child = new Child($value);
	}
}

class Child
{
	public function __construct( $value ) {
		throw new Exception('Bad Value');
	}
}

try {
	$mytop = new Top($_POST);
} catch( Exception $e ) {
	echo 'You had an exception';
}


In the case above, it'd obviously be trivial to change, but with complex class relationships, and factory methods as my particular case involves, it gets cumbersome.  Thus, it seems that instantiation of a class that might have some parts of it not successfully work out, which could be totally acceptable, would cause the Top class to not be created at all, and force the $mytop = new Top($_POST); to be done outside the try/catch, with subsequent checks done inside a try/catch.


---
Hans Zaunere
President, Founder
New York PHP
<A HREF="http://www.nyphp.org">http://www.nyphp.org</A>



</PRE>

<!--endarticle-->
    <HR>
    <P><UL>
        <!--threads-->
	<LI>Previous message: <A HREF="013596.html">[nycphp-talk] PHP 5 Throwing Exceptions in Constructors
</A></li>
	<LI>Next message: <A HREF="013600.html">[nycphp-talk] PHP 5 Throwing Exceptions in Constructors
</A></li>
         <LI> <B>Messages sorted by:</B> 
              <a href="date.html#13599">[ date ]</a>
              <a href="thread.html#13599">[ thread ]</a>
              <a href="subject.html#13599">[ subject ]</a>
              <a href="author.html#13599">[ author ]</a>
         </LI>
       </UL>

<hr>
<a href="http://lists.nyphp.org/mailman/listinfo/talk">More information about the talk
mailing list</a><br>
</body></html>
